### PR TITLE
Support calling EVP_DigestUpdate instead of EVP_Digest[Sign|Verify]Update

### LINKED ISSUE #10114: EVP_DigestUpdate crashes with dgst command-line
Hey OpenSSL team, our [Triggerflow](https://gitlab.com/nisec/openssl-triggerflow-ci/pipelines) tool detected a [problem](https://gitlab.com/nisec/openssl-triggerflow-ci/-/jobs/310739852) during some of our testing in master.

Further inspection it seems that `EVP_DigestUpdate` crashes when using the dgst command-line tool to generate DSA signatures. For example,
```
openssl dgst -sha512 -sign dsa.pem -out lsb-release.sig data
```

I believe the issue is in a NULL pointer indirection in `EVP_DigestUpdate`. 
```
legacy:
    return ctx->update(ctx, data, count);
```

### gdb output
```
(gdb) p *ctx
$1 = {reqdigest = 0x7ffff785c520 <sha512_md>, digest = 0x0, engine = 0x0, flags = 0, md_data = 0x0, pctx = 0x55555584c830, update = 0x0, provctx = 0x0, fetched_digest = 0x0}
```

We appreciate if you could take a look into this.

### PR DESCRIPTION
**Pull request by mattcaswell**
Prior to OpenSSL 3.0 EVP_Digest[Sign|Verify|Update were just macros for
EVP_DigestUpdate. They are now separate functions. Unfortunately some
code assumes that EVP_Digest[Sign|Verify]Update is interchangeable with
EVP_DigestUpdate. For example the dgst app uses an MD bio which always
calls EVP_DigestUpdate(). However the dgst app supports signing instead
of digesting and may initialise with EVP_DigestSignInit_ex() instead of
just EVP_DigestInit().

We now detect these differences and redirect to the correct function
where appropriate.

Fixes #10114

A secondary issue related to #10114 is that the implementation of the MD BIO used internal EVP_MD_CTX knowledge that hadn't been updated with knowledge about providers. There's really no reason for that code to use internal knowledge so instead we use the public APIs.

### CODE REVIEW

File: test/evp_extra_test.c [hunk 1]
```diff
@@ -420,24 +490,55 @@ static int test_EVP_Enveloped(void)
...
+ * Test 3: Use an MD BIO to do the Update calls instead (DSA)
+ */
+static int test_EVP_DigestSignInit(int tst)
 {
     int ret = 0;
     EVP_PKEY *pkey = NULL;
     unsigned char *sig = NULL;
     size_t sig_len = 0;
     EVP_MD_CTX *md_ctx, *md_ctx_verify = NULL;
+    BIO *mdbio = NULL, *membio = NULL;
```
Comment (romen) [inline test/evp_extra_test.c]: Travis failure is relevant : md_ctx above should also be initialized to NULL

**Comment by mattcaswell** [comment]
Fixup pushed addressing the travis failure. The compiler warning was actually bogus - but fixed up anyway. Please take another look.

File: test/evp_extra_test.c [hunk 2]
```diff
@@ -447,19 +548,38 @@ static int test_EVP_DigestSignInit(void)
...
  out:
-    EVP_MD_CTX_free(md_ctx);
-    EVP_MD_CTX_free(md_ctx_verify);
+    if (tst >= 2) {
+        BIO_free(membio);
+        BIO_free(mdbio);
+    } else {
+        EVP_MD_CTX_free(md_ctx);
+        EVP_MD_CTX_free(md_ctx_verify);
+    }
```
Comment (levitte) [inline test/evp_extra_test.c]: Since all these variables are now properly initialized with NULL, there should be no more need to check `tst` here.

Comment (mattcaswell) [reply test/evp_extra_test.c]: Err, no. We still need the check. Depending on which test we are running we have either created the ctx values ourselves or got them out of the BIO. If they are out of the BIO then we are not the owner so we should not free them.

Comment (levitte) [reply test/evp_extra_test.c]: Ah. So the trade-off is to have a couple of extra variables to hold the pointers this function owns, and any time you do allocate a context, you do something like this:

``` C
    if (!TEST_ptr(md_ctx = a_md_ctx = EVP_MD_CTX_new())
```

... and then just free `a_md_ctx`

The benefit that method is avoiding having to remember what condition what variable was given what value ; it's less error prone.

Comment (mattcaswell) [reply test/evp_extra_test.c]: Yes, it can certainly be done that way. Fixup commit pushed.

File: test/evp_extra_test.c [hunk 3]
```diff
@@ -576,10 +577,9 @@ static int test_EVP_DigestSignInit(int tst)
     if (tst >= 2) {
```
Comment (levitte) [inline test/evp_extra_test.c]: And you're still keeping this condition...  why?

**Comment by mattcaswell** [comment]
Pushed.Thanks.

**Pull request closed without merging**